data/transactions/logic/eval.go

```diff
@@ -2526,22 +2527,44 @@ func opEd25519verify(cx *EvalContext) {
 }
 
 func opLoad(cx *EvalContext) {
-	gindex := int(uint(cx.program[cx.pc+1]))
-	cx.stack = append(cx.stack, cx.scratch[gindex])
+	n := cx.program[cx.pc+1]
+	cx.stack = append(cx.stack, cx.scratch[n])
+}
+
+func opLoads(cx *EvalContext) {
+	last := len(cx.stack) - 1
+	n := cx.stack[last].Uint
+	if n >= uint64(len(cx.scratch)) {
+		cx.err = fmt.Errorf("invalid Scratch index %d", n)
+		return
+	}
+	cx.stack[last] = cx.scratch[n]
 }
 
 func opStore(cx *EvalContext) {
-	gindex := int(uint(cx.program[cx.pc+1]))
+	n := cx.program[cx.pc+1]
 	last := len(cx.stack) - 1
-	cx.scratch[gindex] = cx.stack[last]
+	cx.scratch[n] = cx.stack[last]
 	cx.stack = cx.stack[:last]
 }
 
-func opGloadImpl(cx *EvalContext, groupIdx int, scratchIdx int, opName string) (scratchValue stackValue, err error) {
-	if groupIdx >= len(cx.TxnGroup) {
+func opStores(cx *EvalContext) {
+	last := len(cx.stack) - 1
+	prev := last - 1
```

Comment:
I would prefer if we swap the order of arguments here. In opcodes which set a value (e.g. `app_global_put`), we normally make the first arg the location and the second the value. `<index>; <value>; stores` seems more intuitive given the behavior of our other opcodes.

Reply:
I don't feel incredibly strongly about it, but I'll tell you my reasoning and see if you find it convincing.

In the other cases where we have instructions with one version that takes a stack value instead of an immediate, we tacked it on to the "end", that is, top of stack. substring3, extract3, gtxnsas are the only "real" examples of this though, as they are the only ones where the non-stack version had stack arguments already, and so order matters.

What do you think?  I'm truly willing to go either way, whichever you think will give a bit more intuition to users.

Reply:
From my point of view, `stores` is the general purpose version of the opcode, and `store <index>` in an optimized version that can be used when the index is a compile-time constant. So I think it's ok if the argument order of these two versions differ. I'm more interested in having `stores` make sense on its own.

We already have a similar situation with `gtxnas` and `gtxnsas`, since their order is `<array-index>; gtxnas <txn-index> <field>` vs `<txn-index>; <array-index>; gtxnsas <txn-field>`. The order of arguments isn't consistent there, but I think each of the opcodes makes sense on its own.

Reply:
ok, good enough for me.  I'll swap them.